docs/tutorial/streamer_design.md

```diff
@@ -1,29 +1,398 @@
+  for(tb_n_2 = 0; tb_n_2 < temporal_bound[n-2]; tb_n_2++)
+   …
+    for(tb_0 = 0; tb_0 < temporal_bound[0]; tb_0++)
+     parfor(sb_m_1 = 0; sb_m_1 < spatial_bound[m-1]; sb_m_1++)
+      parfor(sb_m_1 = 0; sb_m_2 < spatial_bound[m-2]; sb_m_1++)
+       …
+        parfor(sb_0 = 0; sb_0 < spatial_bound[0]; sb_0++)
+
+          temporal_address = tb_n_1*temporal_stride[n-1] + tb_n_2*temporal_stride[n-2] + ... +
+                             tb_1*temporal_stride[1] + tb_1*temporal_stride[0]
+                        
+          spatial_address = sb_n_1*spatial_stride[n-1] + sb_n_2*spatial_stride[n-2] + ... +
+                            sb_1*spatial_stride[1] + sb_0*spatial_stride[0]
+
+          target_address = base_address + temporal_address + spatial_address
+```
+
+Where `temporal_bound[*]` pertains to the bounds of the temporal loops and `tb_*` pertains to the temporal indices. The `spatial_bound[*]` peratins to the bounds of the spatial loops and `sb_*` pertains to the spatial indices. Wroup them according to `temporal_address` and `spatial_address` and add them to the `base_address` which results in the `target_address`.
+
+The affine address generation is the working principle of the SNAX streamer. With this, we can flexibily access data in various places of the memory. 
+
+# Streamer Microarchitecture
+
+The figure below shows a more detailed architecture of the streamer. The **(1) streamer** sits between the TCDM interconnect and the accelerator. There is also a **(2) streamer wrapper** to re-wire the Chisel-generated signals. More details of the wrappers are in [Connect The Shell](./connect_shell.md) section.
+
+![image](https://github.com/KULeuven-MICAS/snitch_cluster/assets/26665295/37900d11-504d-4659-ba9a-aa7efe589975)
+
+## Streamer Interfaces
+
+### **(3) TCDM interface** 
+
+The **TCDM interface** has request and response channels and also use a decoupled interface. The table below describe the details of the request and response channels:
+
+| signal name        | description                                |
+| :-----------------:| :----------------------------------------: |
+| tcdm_req_addr_o    | Target address in memory.                  |
+| tcdm_req_data_o    | Data to write in memory.                   |
+| tcdm_req_write_o   | Write enable signal. 1 = write, 0 = read.  |
+| tcdm_req_strb_o    | Byte strobe.                               |
+| tcdm_req_core_id_o | Core ID. Currently unused.                 |
+| tcdm_req_is_core_o | Request from core. core. Currently unused. |
+| tcdm_req_valid_o   | Request channel is valid.                  |
+| tcdm_req_ready_i   | Request channel is ready.                  |
+| tcdm_rsp_data_i    | Response data coming from memory.          |
+| tcdm_rsp_valid_i   | Response data is valid.                    |
+
+!!! note
+
+  Some of the signals are currently unused by the streamer. However, we needed to comply with the TCDM IP of the Snitch platform.
```

Comment:
Now I'm wondering why they are not used? :smile:

Reply:
Good question! Technically some of there weren't even used in the original Snitch repo ^^;

I think the others were meant for transactions between clusters. But locally not. For example:

https://github.com/KULeuven-MICAS/snitch_cluster/blob/26427629e2dec6d8911aa06f3359feca64ba085f/hw/snitch_cluster/src/snitch_cluster.sv#L661-L663

I think it's just an added feature if you want to use it. So for now, we aren't using it but we could later!